stats: Remember recent lookups and display them in an admin endpoint

jmarantz opened the pull request:
Description: Adds an admin endpoint to show recent lookups. This blocks #4980 .
Risk Level: medium -- this does add overhead when creating new StatNames.
Testing: //test/...
Docs Changes: yes, in the PR.
Release Notes: in the PR.
Fixes: https://github.com/envoyproxy/envoy/issues/8035

jmarantz commented:
This is ready for a round of review, though I realize I need to work a little more on the stats.md doc.

jmarantz commented:
I have an un-pushed change on top of this that adds logging. The extra code is small (~ 300 lines) but the implication is risk of spamming logs.

Currently I have this set to dump the remembered (10) recent lookups for the last 5 minutes, at most once every 5 minutes. I am wondering if that should be flag-controlled and if therefore we should review this as is, and then let the logging addition be a follow-up.

I'm also thinking that a config API to add builtin stat-names would make any reports in those logs actionable by Envoy users without having to make code changes. Maybe that should all be integrated into this one PR? WDTY?

mattklein123 commented:
@jmarantz do you want to review this in parallel to https://github.com/envoyproxy/envoy/pull/8243? Or wait until that one is done?

jmarantz commented:
#8243 should probably go in first though you could review in parallel.

mattklein123 commented on source/common/stats/symbol_table_impl.h:
```diff
@@ -173,6 +180,9 @@ class SymbolTableImpl : public SymbolTable {
   // This must be held during both encode() and free().
   mutable Thread::MutexBasicLockable lock_;
 
+  // This must be held while updateing stat_name_sets_.
```
typo updateing

mattklein123 commented on source/common/stats/recent_lookups.h:
```diff
@@ -0,0 +1,63 @@
+namespace Envoy {
+namespace Stats {
+
+// Remembers the last 'Capacity' items passed to lookup().
```
I have some high level questions before doing a deep-dive review. 
1) Does time really matter? Would it be better to actually keep a map of name to count with some bounded size with dropping the least counted stats?
2) With the stat set integration, would it be worth it to somehow track the set name and its counts differently since it's a different mutex? Also, how does this relate to the built-in/dynamic stuff in the other PR.
3) I really think we need some type of stat that tracks overall stat name lookup in the slow path, otherwise how would someone know that something might be wrong? Could we have some type of global lookup counter that could be monitored?

jmarantz replied:
1. I think time matters in that I don't want to see any of the initial burst of lookups which have to occur on startup, not on the hot path.

2. I am tracking them separately; everyone has separate RecentLookups deques. But they could be displayed separately rather than interleaved together in the admin console.

3. I don't think we know exactly what path we are in. It could be a one-time registration of some interesting new names in response to an xDS event, or it could be in response to a request. From inside the stats APIs I don't think I know (do I?). Anyway the RecentLookups structure does count lookups and we could emit those in admin /stats. I think I was planning to do that and didn't get to it. I can add it. Are you thinking we should also track nanoseconds spent and report that too? That might be interesting.

jmarantz replied:
Oh but I guess the absolute time doesn't matter; I could solve that with a sequence number instead, and thus avoid having to pass in the TimeSource and make system calls. But if we're going to measure the real-time consumed we'll be collecting a timestamp anyway.

mattklein123 replied:
> Oh but I guess the absolute time doesn't matter; I could solve that with a sequence number instead, and thus avoid having to pass in the TimeSource and make system calls. But if we're going to measure the real-time consumed we'll be collecting a timestamp anyway.

TBH I'm not sure I fully grok what is going on here which may be an indication we should try to simplify. 

I think as a user what I really want to know is whether some stat name keeps getting looked up over and over again. I don't really care about single lookups, and I'm not sure I care about time either. So if we kept a map of name to count of lookups and just output the raw count from max, dropping off at a max table length, wouldn't this give me enough visiblity?

For the stat itself, I was thinking about just a raw counter of lookups under lock, which again we could roughly correlate to a lot of steady state lookups vs. just a small burst at startup?

jmarantz replied:
Actually in some ways the pending log change simplifies things. I gave up on making RecentLookups templated and just hold a string, to avoid locking issues during logging. I think I'll go ahead and push that; we can always back it out or decide to make a command-line option.

map<name,count> works functionally but how would you want to evict?  I didn't want to keep all the names in the map because that would erase the memory benefit from this work; I wanted to have bounded memory use.

I also thought of putting in a small lru-cache impl but settled on a deque as a cheaper good-enough approximation of that.

At one point in this PR I did collate the deque and print counts. I could bring that back; I'd just collate while printing. In that case i might increase the deque size from 10 to 100 so there's a more to correlate.

The reason I thought of printing the results with the timestamp is that if the lookups were pretty sparse you could correlate the timestamps in the admin console with the server-start time and if all the lookups happened right after startup or xDS update then you could ignore them.

One thing for sure: I'm not really sure how useful the logging or admin console will be until some traffic goes through. I'm sure we'll want to tweak it once there's some mileage on it.

mattklein123 replied:
What about a bounded max heap and then just sort by count? All of the single lookups would just drop off?

jmarantz replied:
Matt & I discussed on Slack a bit ... possibilities include some kind of LFU (heap + map) or LRU (list + map). I find LRU easier to reason because I don't know how to admit new entries into a pure LFU.

I am concerned about the raw low-level speed of the data structure because this is running under the global lock potentially, and so I added a speed test. I got this about 10ns per lookup overhead for the deque, and about 268ns per lookup for the LRU structures. Note that the LRU is all O(1) complexity but a bunch of pointers & hashes to update on every lookup, so the constant-time overhead is 20x that of the deque.

I added the LRU impl in https://github.com/jmarantz/envoy/blob/remember-recent-lookups-lru/source/common/stats/recent_lookups.h and .cc

I added the speed test in this PR. Nothing here is that well tested yet but that's my first cut.

I'm leaning toward the deque impl in this PR as it reports the same data (items and counts) with the same API, with much less overhead. The drawback of the deque is that some potentially interesting entries will age out. But if the idea is that we should be eliminating all those anyway, it should be fine.

mattklein123 replied:
@jmarantz we should ship and iterate on the dequeue. So will you drop time then and just have items and counts?

jmarantz commented on source/common/stats/recent_lookups.h:
```diff
@@ -0,0 +1,63 @@
+  /**
+   * @return the total number of lookups since tracking began.
+   */
+  uint64_t total() const { return total_; }
```
here's the counter under lock; just need to expose it.

mattklein123 replied:
+1

mattklein123 commented on include/envoy/stats/BUILD:
```diff
@@ -40,6 +40,7 @@ envoy_cc_library(
     name = "symbol_table_interface",
     hdrs = ["symbol_table.h"],
     deps = [
+        "//include/envoy/common:time_interface",
```
not needed anymore?

mattklein123 commented on include/envoy/stats/symbol_table.h:
```diff
@@ -5,6 +5,7 @@
 #include <vector>
 
 #include "envoy/common/pure.h"
+#include "envoy/common/time.h"
```
del?

mattklein123 commented on include/envoy/stats/symbol_table.h:
```diff
@@ -183,6 +197,22 @@ class SymbolTable {
   virtual StoragePtr encode(absl::string_view name) PURE;
+
+  /**
+   * Enables trackRecentLookups to also track lookups that occur in
```
Would it be better to have the stat name set be created from the symbol table such that a reference can be embedded and this handled automatically? Is it this way because of the extra memory that would require?

mattklein123 replied:
> one option I guess is to have SymbolTable::makeSet(). That could also work, but would be an API change for all the existing StatNameSets. If you think that is a lot better we can change it though.

IMO that would be cleaner, yes, if you don't mind.

jmarantz replied:
don't mind; will do that a bit later. will be non-controversial but a few files hit.

jmarantz replied:
ok; done...this added a bunch of files to the PR and we could potentially break that out as a separate one if desired.

jmarantz replied:
resolving assuming the PR is not too big to review now.

mattklein123 commented on source/common/stats/recent_lookups.h:
```diff
@@ -0,0 +1,42 @@
+  using IterFn = std::function<void(absl::string_view, uint64_t)>;
+
+  /**
+   * Calls fn(item, timestamp) for each of the remembered lookups.
```
timestamp -> count?

mattklein123 commented on source/common/stats/recent_lookups.cc:
```diff
@@ -0,0 +1,40 @@
+void RecentLookups::forEach(IterFn fn) const {
+  absl::flat_hash_map<absl::string_view, uint64_t> counts;
+  for (const std::string& item : queue_) {
+    ++counts[item];
```
I'm a little confused here. Isn't the idea to keep counts of things that we keep looking up? With a capacity of 10 are you hoping to just get lucky if the same thing keeps getting looked up over and over again?

mattklein123 replied:
Sorry stepping back I'm still confused though. AFAICT the implementation of lookup always just pops the back and pushes to the front.

At the time of iterating over the dequeue I see that we are counting the items, but wouldn't it be better to keep a running count of the lookups as I'm not sure that counting at the time of admin fetch is all that useful as that could be done just as well by a script? I realize we are going in circles a bit, but couldn't this just be a parallel LRU and map, where the map is from string -> count/LRU entry, and then if there is no map entry we just pop off the back? If we don't want to do this, can you flesh out the comments in recent_lookups.h with well specified properties of what we are storing, runtime complexity, tradeoffs, etc.? I think that would help the discussion.

jmarantz replied:
I have a branch implementing the LRU: https://github.com/jmarantz/envoy/blob/8cdf5fc83a2fa43213f6e555452c49808e4fe4bc/source/common/stats/recent_lookups.h#L67 and https://github.com/jmarantz/envoy/blob/8cdf5fc83a2fa43213f6e555452c49808e4fe4bc/source/common/stats/recent_lookups.cc#L25

It's just >20x slower than the deque (measured with a microbenchmark) and the main difference at the end of the day is that for a given memory commitment you will see a lot more of the violating stats with the LRU.

The cost is that anyone running an experiment with the real symbol tables will be holding the global mutex for much longer due to the LRU, and thus increasing contention on real traffic.

If the goal is to just find and eliminate all the contending stats then I think the LRU might make the iterations faster by capturing more violations. But I'I think it's not worth it as it makes the experiments more expensive, and we should be able to converge based on data from the deque.

Take for example the http2 stats I found we were looking up by name on every request in https://github.com/envoyproxy/envoy/issues/8324 -- the 'total' counter told us there are 12, but I only list 10 of them because that's the size of the deque. I'm guessing there's one code-change to fix all 12, but if there wasn't we could just do another run to find out the 2 that fell off. In this particular case the LRU wouldn't have mattered because each was looked up just one time per request.

The LRU probably will help narrow things down faster in some cases.

Anyway that's my reasoning; I'm OK if you prefer I go with LRU anyway in which case I'll document the perf concern, or I could add this rationale in a comment and stay with this one.

mattklein123 replied:
What if the lookup accounting is off by default and possibly enabled via admin endpoint and/or CLI flag?

We can have a stat for total lookups, and then if the operator actually wants to debug in depth they could enable the lookup debugging? WDYT?

jmarantz replied:
SGTM will implement later today during code therapy.

jmarantz replied:
qq; is the 'stat' as I have it; just a line printed from `/stats?recentlookups` sufficient? Or do we want it to be intgrated into `/stats`?

mattklein123 replied:
IMO it should be integrated into `/stats` as an actual `server` stat, as this is what I think people would want to monitor?

jmarantz replied:
so probably I want to make a real stat for this, which I can probably inject via lambda or something.

jmarantz replied:
I have made the LRU implementation but I didn't make the total a real stat yet.

I wanted to get a taste-test for how to do it.

My current thinking is to pass in a lambda to RecentLookups to bump the stat, to avoid a circular dependence, since Counter depends on SymbolTable, and SymbolTable depends on RecentLookups, and RecentLookups should thus not be dependent on Counter. Might be able to finesse this with interface/impl breaks but it seems easier to use a lambda. WDYT?

jmarantz commented:
@mattklein123  here are a few loose ends here, such as documenting all the endpoints, but can you do a quick scan to see if you are aligned with the overall structure?

I added endpoints `/stats/recentlookups/clear`, `/stats/recentlookups/enable`, and `/stats/recentlookups/disable` (all POSTs), and changed the query function to simply `/stats/recentlookups`. I changed the capacity to be settable from the C++ API (which makes it easier to test eviction) though I left it as a const 100 in admin.cc.

mattklein123 commented on include/envoy/stats/symbol_table.h:
```diff
@@ -183,6 +221,13 @@ class SymbolTable {
+   *
+   * @param stat_name_set the set.
+   */
+  virtual void forgetSet(StatNameSet& stat_name_set) PURE;
```
Can this be just implicit inside the created stat set so we don't need this at the interface level? It seems like the created implementation of the set can have an explicit handle to the parent and just remove itself?

mattklein123 commented on include/envoy/stats/symbol_table.h:
```diff
@@ -143,10 +146,45 @@ class SymbolTable {
+   *
+   * @param iter the function to call for every recent item.
+   */
+  virtual uint64_t getRecentLookups(const RecentLookupsFn& iter) PURE;
```
nit: can this be const?

jmarantz commented:
RE splitting: I think I can split this in 3: 
 * RecentLookups and its unit tests
 * StatNameSet creation mechanism
 * Pull it all together (this PR)

The first 2 would be independent of one another.

I'd prefer to leave the `forgetSet` abstraction if that's OK as that is actually virtual, so that the StatNameSet abstraction works with FakeSymbolTableImpl. That could be unvirtualized and inlined once I get rid of FakeSymbolTableImpl.

I can make the stuff const of course, though I might need mutable mutexes.

mattklein123 commented:
> I'd prefer to leave the forgetSet abstraction if that's OK as that is actually virtual, so that the StatNameSet abstraction works with FakeSymbolTableImpl. That could be unvirtualized and inlined once I get rid of FakeSymbolTableImpl.

OK sounds good.

jmarantz commented:
blocked on https://github.com/envoyproxy/envoy/pull/8369 and https://github.com/envoyproxy/envoy/pull/8376

jmarantz commented on source/common/stats/symbol_table_impl.h:
```diff
@@ -176,6 +181,9 @@ class SymbolTableImpl : public SymbolTable {
   mutable Thread::MutexBasicLockable lock_;
 
+  // This must be held while updating stat_name_sets_.
+  mutable Thread::MutexBasicLockable stat_name_set_mutex_;
```
Note we do need a separate mutex for this or we get deadlock detection triggered by absl when calling getDynamic().

mattklein123 commented on source/server/http/admin.cc:
```diff
@@ -691,7 +693,47 @@ Http::Code AdminImpl::handlerResetCounters(absl::string_view, Http::HeaderMap&,
+Http::Code AdminImpl::handlerStatsRecentLookups(absl::string_view, Http::HeaderMap&,
+                                                Buffer::Instance& response, AdminStream&) {
+  Stats::SymbolTable& symbol_table = server_.stats().symbolTable();
+  /*response_headers.insertContentType().value().setReference(
```
?

jmarantz replied:
removed.

mattklein123 commented on docs/root/operations/admin.rst:
```diff
@@ -354,6 +354,22 @@ modify different aspects of the server:
   Envoy has updated (counters incremented at least once, gauges changed at least once,
   and histograms added to at least once)
 
+  .. http:get:: /stats?recentlookups
```
I don't think this path is right, also can you doc all the paths? Also can you add a release note?

jmarantz replied:
I'm still tweaking the doc a bit. I'll re-add 'waiting' tag until it's ready.

jmarantz replied:
ok; doc fixed, release note added.

mattklein123 commented on source/server/http/admin.cc:
```diff
@@ -691,7 +693,47 @@ Http::Code AdminImpl::handlerResetCounters(absl::string_view, Http::HeaderMap&,
+  } else {
+    response.add("   Count Lookup\n");
+  }
+  response.add(absl::StrCat(table, "\ntotal: ", total, "\n"));
```
We still want to expose this as a server stat, right? I think that is the main operational thing that people will monitor?

jmarantz replied:
done.

jmarantz commented:
/retest

mattklein123 commented on docs/root/operations/admin.rst:
```diff
@@ -354,6 +354,55 @@ modify different aspects of the server:
+  but in response to user requests on high core-count machines, this
+  can cause performance issues due to mutex contention.
+
+  This option requires Envoy to be started with `use-fake-symbol-table 0`.
```
nit: I think you can use an `:option` ref link here.

jmarantz replied:
done; thanks.

jmarantz replied:
Actually I backed this out, as `use-fake-symbol-table` is not docced, as it was meant to be temporary.

mattklein123 commented on docs/root/operations/admin.rst:
```diff
@@ -354,6 +354,55 @@ modify different aspects of the server:
+  .. http:post:: /stats/recentlookups/clear
+
+  Clears all outstanding lookups and counts. If called when recent lookup
+  collection is enabled, this clears all the, but collection
```
typo

jmarantz replied:
fixed, and I simplified this, as I don't think it adds anything to the paragraph to discuss the two different cases.

commit coming.
